src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -68,7 +74,7 @@ public static int TrailingZeroCount(uint value)
                 // Using deBruijn sequence, k=2, n=5 (2^5=32) : 0b_0000_0111_0111_1100_1011_0101_0011_0001u
                 ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
                 // long -> IntPtr cast on 32-bit platforms is expensive - it does overflow checks not needed here
-                (IntPtr)(int)(((uint)((value & -value) * 0x077CB531u)) >> 27)); // shift over long also expensive on 32-bit
+                (IntPtr)(((value & (uint)-value) * 0x077CB531u) >> 27)); // shift over long also expensive on 32-bit
```

Comment:
Simplified, per @tannergooding suggestion in https://github.com/dotnet/coreclr/pull/22497
This amounts to a `(IntPtr)(uint …)` conversion. 
Is there any overhead there - should I be casting `(IntPtr)(int)(uint …)` instead?

Reply:
`(IntPtr)(uint …)` conversion does bounds checks on 32-bit platforms.

It would be useful to look at the generated native code for this and `Log2Fallback` on both x86 and x64 to make sure that it is what we expected it to be. The inefficiencies introduced by signed/unsigned casts are very easy to miss.

Reply:
Fixed. Explicit conversion path through `int`, ie `(IntPtr)(int)(uint …)`.

Reply:
Here's the code & MSIL (Release) for `Log2SoftwareFallback`
```csharp
if (value == 0)
    return 32;

return Unsafe.AddByteOffset(
    ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
    (IntPtr)(int)(((value & (uint)-value) * 0x077CB531u) >> 27));
```
```csharp
// if (value == 0)
IL_0000 ldarg.0
IL_0001 brtrue.s  IL_0006
// return 32;
IL_0003 ldc.i4.s  32
IL_0005 ret
// return Unsafe.AddByteOffset(
// ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn),
// (IntPtr)(int)(((value & (uint)-value) * 0x077CB531u) >> 27));
IL_0006 call      System.ReadOnlySpan`1<System.Byte> SourceCode.Clay.Buffers.Bench.TrailingZeroBench::get_s_TrailingZeroCountDeBruijn()
IL_000B call      System.Byte& System.Runtime.InteropServices.MemoryMarshal::GetReference<System.Byte>(System.ReadOnlySpan`1<System.Byte>)
IL_0010 ldarg.0
IL_0011 ldarg.0
IL_0012 conv.u8
IL_0013 neg
IL_0014 conv.u4
IL_0015 and
IL_0016 ldc.i4    125613361
IL_001B mul
IL_001C ldc.i4.s  27
IL_001E shr.un
IL_001F call      System.IntPtr System.IntPtr::op_Explicit(System.Int32)
IL_0024 call      System.Byte& System.Runtime.CompilerServices.Unsafe::AddByteOffset<System.Byte>(System.Byte&,System.IntPtr)
IL_0029 ldind.u1
IL_002A ret
```

Reply:
And for `Log2SoftwareFallback`:
```csharp
value |= value >> 01;
value |= value >> 02;
value |= value >> 04;
value |= value >> 08;
value |= value >> 16;

return Unsafe.AddByteOffset(
    ref MemoryMarshal.GetReference(s_Log2DeBruijn),
    (IntPtr)(int)((value * 0x07C4ACDDu) >> 27));

// value |= value >> 01;
IL_0000 ldarg.0
IL_0001 ldarg.0
IL_0002 ldc.i4.1
IL_0003 shr.un
IL_0004 or
IL_0005 starg.s   value
// value |= value >> 02;
IL_0007 ldarg.0
IL_0008 ldarg.0
IL_0009 ldc.i4.2
IL_000A shr.un
IL_000B or
IL_000C starg.s   value
// value |= value >> 04;
IL_000E ldarg.0
IL_000F ldarg.0
IL_0010 ldc.i4.4
IL_0011 shr.un
IL_0012 or
IL_0013 starg.s   value
// value |= value >> 08;
IL_0015 ldarg.0
IL_0016 ldarg.0
IL_0017 ldc.i4.8
IL_0018 shr.un
IL_0019 or
IL_001A starg.s   value
// value |= value >> 16;
IL_001C ldarg.0
IL_001D ldarg.0
IL_001E ldc.i4.s  16
IL_0020 shr.un
IL_0021 or
IL_0022 starg.s   value
// return Unsafe.AddByteOffset(
// ref MemoryMarshal.GetReference(s_Log2DeBruijn),
// (IntPtr)(int)((value * 0x07C4ACDDu) >> 27));
IL_0024 call      System.ReadOnlySpan`1<System.Byte> SourceCode.Clay.Buffers.Bench.TrailingZeroBench::get_s_Log2DeBruijn()
IL_0029 call      System.Byte& System.Runtime.InteropServices.MemoryMarshal::GetReference<System.Byte>(System.ReadOnlySpan`1<System.Byte>)
IL_002E ldarg.0
IL_002F ldc.i4    130329821
IL_0034 mul
IL_0035 ldc.i4.s  27
IL_0037 shr.un
IL_0038 call      System.IntPtr System.IntPtr::op_Explicit(System.Int32)
IL_003D call      System.Byte& System.Runtime.CompilerServices.Unsafe::AddByteOffset<System.Byte>(System.Byte&,System.IntPtr)
IL_0042 ldind.u1
IL_0043 ret
```

Reply:
`IL_0012 conv.u8` is unnecessary. `-value` may need to be `-(int)value`.

Reply:
Or save the cast by changing the core API to take `int`, and the wrapper API to take `uint`.

Reply:
Log2SoftwareFallback looks ok.

Reply:
Nice suggestion. 
I had to change it slightly to get units to pass, but it's tighter now:
```csharp
...
(IntPtr)(int)(((value & (uint)-(int)value) * 0x077CB531u) >> 27)

IL_0006 call      System.ReadOnlySpan`1<System.Byte> SourceCode.Clay.Buffers.Bench.TrailingZeroBench::get_s_TrailingZeroCountDeBruijn()
IL_000B call      System.Byte& System.Runtime.InteropServices.MemoryMarshal::GetReference<System.Byte>(System.ReadOnlySpan`1<System.Byte>)
IL_0010 ldarg.0
IL_0011 ldarg.0
IL_0012 neg
IL_0013 and
IL_0014 ldc.i4    125613361
IL_0019 mul
IL_001A ldc.i4.s  27
IL_001C shr.un
IL_001D call      System.IntPtr System.IntPtr::op_Explicit(System.Int32)
IL_0022 call      System.Byte& System.Runtime.CompilerServices.Unsafe::AddByteOffset<System.Byte>(System.Byte&,System.IntPtr)
IL_0027 ldind.u1
IL_0028 ret
```